pkg/controller/egressgateway/egressgateway_controller.go

```diff
@@ -0,0 +1,666 @@
+		} else {
+			if len(egw.Spec.Template.Metadata.Labels) > 0 {
+				egw.Spec.Template.Metadata.Labels["projectcalico.org/egw"] = egw.Name
+			} else {
+				egw.Spec.Template.Metadata.Labels = defLabel
+			}
+		}
+	}
+
+	// If affinity isn't specified by the user, default pod anti affinity is added so that 2 EGW pods aren't scheduled in
+	// the same node. If the provider is AKS, set the node affinity so that pods don't run on virutal-nodes.
+	defAffinity := &v1.Affinity{}
+	defAffinity.PodAntiAffinity = &v1.PodAntiAffinity{
+		PreferredDuringSchedulingIgnoredDuringExecution: []v1.WeightedPodAffinityTerm{
+			{
+				Weight: 1,
+				PodAffinityTerm: v1.PodAffinityTerm{
+					LabelSelector: &metav1.LabelSelector{
+						MatchLabels: egw.Spec.Template.Metadata.Labels,
+					},
+					TopologyKey: "topology.kubernetes.io/zone",
+				},
+			},
+		},
+	}
+	switch installation.KubernetesProvider {
+	case operatorv1.ProviderAKS:
+		defAffinity.NodeAffinity = &v1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+				NodeSelectorTerms: []v1.NodeSelectorTerm{{
+					MatchExpressions: []v1.NodeSelectorRequirement{
+						{
+							Key:      "type",
+							Operator: v1.NodeSelectorOpNotIn,
+							Values:   []string{"virtual-node"},
+						},
+						{
+							Key:      "kubernetes.azure.com/cluster",
+							Operator: v1.NodeSelectorOpExists,
+						},
+					},
+				}},
+			},
+		}
+	default:
+		defAffinity.NodeAffinity = nil
+	}
+
+	if egw.Spec.Template.Spec == nil {
+		egw.Spec.Template.Spec = &operatorv1.EgressGatewayDeploymentPodSpec{Affinity: defAffinity}
+	} else if egw.Spec.Template.Spec.Affinity == nil {
+		egw.Spec.Template.Spec.Affinity = defAffinity
+	}
+}
+
+func validateIPPool(ctx context.Context, cli client.Client, ipPool operatorv1.EgressGatewayIPPool, awsNativeIP operatorv1.NativeIP) error {
+	if ipPool.Name != "" {
+		instance := &crdv1.IPPool{}
+		key := types.NamespacedName{Name: ipPool.Name}
+		err := cli.Get(ctx, key, instance)
+		if err != nil {
+			return err
+		}
+		if ipPool.CIDR != "" {
+			if instance.Spec.CIDR != ipPool.CIDR {
+				return fmt.Errorf("IPPool CIDR does not match with name")
+			}
+		}
+		if awsNativeIP == operatorv1.NativeIPEnabled && instance.Spec.AWSSubnetID == "" {
+			return fmt.Errorf("AWS subnet ID must be set when NativeIP is enabled")
+		}
+		return nil
+	}
+	if ipPool.CIDR != "" {
+		instance := &crdv1.IPPoolList{}
+		err := cli.List(ctx, instance)
+		if err != nil {
+			return err
+		}
+		for _, item := range instance.Items {
+			if item.Spec.CIDR == ipPool.CIDR {
+				if awsNativeIP == operatorv1.NativeIPEnabled && item.Spec.AWSSubnetID == "" {
+					return fmt.Errorf("AWS subnet ID must be set when NativeIP is enabled")
+				}
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("IPPool matching CIDR = %s not present", ipPool.CIDR)
+}
+
+func getCumulativeEgressGatewayStatus(egws []operatorv1.EgressGateway) (bool, *operatorv1.EgressGateway) {
```

Comment:
The func name does not align with what it is doing. It is returning an EgressGateway, not a status. 

And the boolean seems redundant as it is always equal to (ewg !=nil)

Should it be named getUnreadyEgressGateway() or something like that?